Introduce BlindedPathParams

shaavan opened the pull request:
1. Added a new struct called `BlindedPathParams` to encapsulate multiple
   properties that define a blinded path.
2. These properties include the number of paths, the compactness of the path,
   and the number of hops.
3. This struct provides a systematic way for users to specify these properties
   before creating a blinded path, offering more flexibility and control.
4. In this version, the struct supports two properties:
   - The number of paths to generate.
   - Whether the path should be compact or not.

shaavan commented:
Updated from [pr3246.01](https://github.com/shaavan/rust-lightning/commits/pr3246.01) to [pr3246.02](https://github.com/shaavan/rust-lightning/commits/pr3246.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.01..pr3246.02)):

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
Updated from [pr3246.02](https://github.com/shaavan/rust-lightning/commits/pr3246.02) to [pr3246.03](https://github.com/shaavan/rust-lightning/commits/pr3246.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.02..pr3246.03)):

Changes:
1. Fix docs.
2. Introduce tests for offer and refund with no blinded paths.

shaavan commented:
Updated from [pr3246.02](https://github.com/shaavan/rust-lightning/commits/pr3246.02) to [pr3246.03](https://github.com/shaavan/rust-lightning/commits/pr3246.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.02..pr3246.03)):

Changes:
1. Fix docs.
2. Introduce tests for offer and refund with no blinded paths.

shaavan commented:
Updated from [pr3246.03](https://github.com/shaavan/rust-lightning/commits/pr3246.03) to [pr3246.04](https://github.com/shaavan/rust-lightning/commits/pr3246.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.03..pr3246.04)):

Changes:
1. Rebase on main, to resolve merge conflicts.
2. Fix CI.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -417,6 +417,22 @@ pub struct PendingOnionMessage<T: OnionMessageContents> {
 	pub reply_path: Option<BlindedMessagePath>,
 }
 
+/// A placeholder value used for paths within the codebase.
+///
+/// PATHS_PLACEHOLDER` is temporarily used as a default value in situations
+/// where a path index is required but has not yet been assigned or initialized.
+pub const PATHS_PLACEHOLDER: usize = 3;
```
Rather than using this globally, we should use a different constant that makes sense wherever this is required. For example, this is used in `handle_onion_message_response` for a reply path even though only one path can ever be used.

shaavan replied:
Makes sense!
Removed the global variable and instead introduced a default constructor for function using the default number of paths in [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -417,6 +417,22 @@ pub struct PendingOnionMessage<T: OnionMessageContents> {
 
+/// A placeholder value used for paths within the codebase.
+///
+/// PATHS_PLACEHOLDER` is temporarily used as a default value in situations
```
Doesn't look like this is temporary.

shaavan replied:
Removed the global variable in [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06)
Thanks for the suggestion!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -487,7 +487,13 @@ pub trait MessageRouter {
 			.collect();
-		self.create_blinded_paths(recipient, context, peers, secp_ctx)
+
+		// This parameter is a placeholder. This function is removed in the subsequent commits.
```
Doesn't look like it was removed.

shaavan replied:
Removed the comment, and the function in [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06)
Thanks for the pointer!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8840,17 +8849,15 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
-		};
+
+			builder = builder.path(path);
+		}
```
Could we use the `match` syntax instead? That way our variables don't need to be `mut`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8914,15 +8922,19 @@ macro_rules! create_refund_builder { ($self: ident, $builder: ty) => {
-		let builder = RefundBuilder::deriving_payer_id(
+		let mut builder = RefundBuilder::deriving_payer_id(
 			node_id, expanded_key, nonce, secp_ctx, amount_msats, payment_id
 		)?
 			.chain_hash($self.chain_hash)
-			.absolute_expiry(absolute_expiry)
-			.path(path);
+			.absolute_expiry(absolute_expiry);
+
+		if let Some(params) = params {
+			let path = $self.create_blinded_paths(params, context)
+				.and_then(|paths| paths.into_iter().next().ok_or(()))
+				.map_err(|_| Bolt12SemanticError::MissingPaths)?;
+
+			builder = builder.path(path);
+		};
```
Likewise.

shaavan replied:
Done! [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06)
Thanks for the suggestion!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9357,6 +9347,7 @@ where
 	///
 	/// Errors if the `MessageRouter` errors.
+	#[allow(unused)]
 	fn create_compact_blinded_paths(&self, context: OffersContext) -> Result<Vec<BlindedMessagePath>, ()> {
```
Why not remove this instead?

shaavan replied:
Remove the function instead of using #[allow(unused)] in [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06)
Thanks for the pointer!

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -391,6 +391,53 @@ fn prefers_more_connected_nodes_in_blinded_paths() {
+
 /// Checks that blinded paths are compact for short-lived offers.
 #[test]
 fn creates_short_lived_offer() {
```
The naming at least for the following tests is no longer relevant, it seems.

shaavan replied:
As far as I understand, with the introduction of `BlindedPathParameters`, it seems like the behavior these tests were checking isn't relevant anymore. Should we think about removing these tests?

jkczyz replied:
Let's wait on the resolution of https://github.com/lightningdevkit/rust-lightning/pull/3246#discussion_r1745858783. We'll ultimately want to test cases allowed by whatever interface is exposed to the user.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -300,8 +300,12 @@ fn prefers_non_tor_nodes_in_blinded_paths() {
 	announce_node_address(charlie, &[alice, bob, david, &nodes[4], &nodes[5]], tor.clone());
 
+	let params = BlindedPathParams {
+		paths: PATHS_PLACEHOLDER,
```
Here and throughout, we shouldn't change the number of paths used in the tests unless we are testing a specific scenario (which we should also do in a new test).

shaavan replied:
Done! Update the file to use the default constructor instead of setting the number of paths using the global constant in [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06)
Thanks a lot for the suggestion!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -432,6 +432,22 @@ pub enum MessageSendInstructions {
+/// This struct allows to specify the properties of a Blinded Path before creating it.
+pub struct BlindedPathParams {
+	/// Number of [`BlindedMessagePath`]s to be created.
+	pub paths: usize,
```
Paths aren't only for QR codes. We use them in reply paths and in messages not transmitted via QR code. The router doesn't know this context.

jkczyz replied:
I guess the `MessageRouter` does get the `MessageContext` now, though that puts the onus on the `MessageRouter` implementation to take the correct action. Maybe that's ultimately desirable? Not sure given you would need to reimplement the logic in `DefaultMessageRouter`.

(Overriding `MessageRouter` is a little annoying in that `ChannelManager` has a `Router` which also requires `MessageRouter`, and `DefaultRouter` also wraps `DefaultMessageRouter`. So it currently requires making a `Router` that wraps `DefaultRouter` and your own `MessageRouter`, implementing the latter by delegating to it as needed. Making the traits independent would make this less cumbersome, if adding another type parameter to `ChannelManager`.)

I think we still want a way for the user to say they don't want any blinded paths in an offer or refund, though.

jkczyz replied:
I'm not sure I'm convinced asking the user re-implement the `MessageRouter` trait is really that trivially. Not only given the above, but for example if they are sure they want to create an offer with one blinded path (perhaps for a QR code) and another offer with many blinded paths (perhaps for BIP 353). Maybe they call a method on their `MessageRouter` that puts it in a different mode before calling `pay_for_offer`? It doesn't seem as intuitive as passing a parameter to `pay_for_offer`.

shaavan replied:
With the [pr3246.07](https://github.com/shaavan/rust-lightning/commits/pr3246.07) I have updated the struct to be simpler. So that maintains the control to MessageRouter to decide how the BlindedPath should look!
So I am marking this as resolved for now!
Thank you so much!

shaavan commented:
Updated from [pr3246.05](https://github.com/shaavan/rust-lightning/commits/pr3246.05) to [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.05..pr3246.06)):
Addressed @jkczyz comments

Changes:
1. Removed the global constant PATHS_PLACEHOLDER, and instead use a default constructor, and local DEFAULT_VALUE.
2. Remove the redundant functions in their appropriate commits.
3. Use `match` to avoid `mut` variables.

shaavan commented:
Updated from [pr3246.06](https://github.com/shaavan/rust-lightning/commits/pr3246.06) to [pr3246.07](https://github.com/shaavan/rust-lightning/commits/pr3246.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.06..pr3246.07)):

Changes:
1. Introduce a new approach using the BlindedPathType enum.
2. The enum allows for the specification of the type of Blinded Path (Compact or Full) that a user can specify to create the desired type of Blinded Path.
3. Update offer_builder and refund_builder so that user can explicitly specify the type of Blinded Path they want to create.
4. Update the Blinded Path creation flow so that only one function flow is responsible for creating both kinds of Blinded Paths.

shaavan commented:
Updated from [pr3246.07](https://github.com/shaavan/rust-lightning/commits/pr3246.07) to [pr3246.08](https://github.com/shaavan/rust-lightning/commits/pr3246.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.07..pr3246.08)):

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
Updated from [pr3246.08](https://github.com/shaavan/rust-lightning/commits/pr3246.08) to [pr3246.09](https://github.com/shaavan/rust-lightning/commits/pr3246.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.08..pr3246.09)):

Changes:
1. Rebase on main, and fix ci.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -444,6 +444,14 @@ pub enum MessageSendInstructions {
 
+/// Represents the types of [`BlindedMessagePath`] that can be created.
+pub enum BlindedPathType {
+	/// Whether the created [`BlindedMessagePath`]s are compact.
+	Compact,
+	/// Whether the created [`BlindedMessagePath`]s are full-length.
+	Full,
```
Let's describe what these mean and why someone would chose one over the other.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9121,16 +9122,32 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
+				let context = MessageContext::Offers(context);
+				let path = $self
+					.create_blinded_paths(context)
+					.and_then(|paths| paths.into_iter().next().ok_or(()))
```
Hmm... @TheBlueMatt I don't think this interface is sufficient to allow someone to specify they want more than one path in the offer, unless we include all the paths returned by the `MessageRouter` instead of just one.

jkczyz replied:
What should we have `DefaultMessageRouter` do?

jkczyz replied:
I suppose `DefaultMessageRouter` could use _both_ `MessageContex` and `BlindedPathType` -- or which `MessageRouter` method is called, if we revert to that state -- to determine if more than one path should be returned. That way we'd allow users to pass `BlindedPathType::Full` to `create_offer_builder` such that they can create an offer with more than one path (e.g., when an offer doesn't need to be in a QR code).

jkczyz replied:
Yeah, seems like either of these approaches should suffice. I'd prefer the two separate method approach, so basically keeping the `MessageRouter` as it was and changing `DefaultMessageRouter::create_compact_blinded_paths` to only return one path.

jkczyz replied:
The point was to allow users creating an offer/refund via `ChannelManager` to specify whether they want no blinded paths, blinded paths that can fit in a QR code, or blinded paths where size constraints is not a limitation.
- Passing `None` for `BlindedPathType` adds support for the first
- Passing `Some(BlindedPathType::Compact)` keeps support for the second
- Passing `Some(BlindedPathType::Full)` and selecting all paths returned by `MessageRouter` adds support for the third

jkczyz replied:
If we change `MessageRouter` to only have one method, then we need to pass in `BlindedPathType`. But then the `peers` parameter will need to be `Vec<MessageForwardNode>` even when used with `BlindedPathType::Full` instead of `Vec<PublicKey>`. So callers could pass in a `MessageForwardNode` with an scid even though it isn't expected, resulting in the last hop being compact.

shaavan replied:
> So callers could pass in a MessageForwardNode with an scid even though it isn't expected, resulting in the last hop being compact.

Makes sense! I have reintroduced the two function flow back in [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11)

jkczyz replied:
> I'm confused why this is a concern, if the `MessageRouter` insists on a specific type of blinded path in general, so what? Ideally the `MessageRouter` gets to pick what it wants to do,

Nothing is preventing a `MessageRouter` from doing so. It would just need access to the channels.

It's just kinda ugly for `DefaultMessageRouter` to need to clear `short_channel_id` depending on how it is called. Plus, `ChannelManager` needs to do additional lookups to get this information just to have it discarded. I'm fine either way, but it doesn't optimize the usual case (i.e., use non-compact paths for reply paths).

> though I think we're not allowed to have a compact introduction point in reply paths? Even in that case, though, ISTM we can communicate the restrictions and if the `MessageRouter` is buggy we can just fail.

Yeah, though it doesn't look like we are enforcing this at the moment.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9121,16 +9122,32 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
-			None => builder,
-			Some(absolute_expiry) => builder.absolute_expiry(absolute_expiry),
+				OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
+					.chain_hash($self.chain_hash)
```
Let's DRY this up by moving this part outside the `match`.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -477,24 +472,12 @@ pub trait MessageRouter {
 		T: secp256k1::Signing + secp256k1::Verification
 	>(
-		&self, recipient: PublicKey, context: MessageContext,
+		&self, recipient: PublicKey, context: MessageContext, blinded_path: BlindedPathType,
```
Discussed offline. Probably better to keep this as two separate methods. But still unresolved as to whether how the number of paths used should be specified. Let's wait until that is resolved before reverting though.

jkczyz replied:
Yeah, let's revert those two commits and move `BlindedPathType` to `channelmanager`, since it is only needed there.

shaavan replied:
Done!
Reintroduced the two-function for blinded path creation in [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11)
Also moved BlindedPathType to channelmanager.
Thanks for the pointer! 🌟

shaavan commented:
Updated from [pr3246.09](https://github.com/shaavan/rust-lightning/commits/pr3246.09) to [pr3246.10](https://github.com/shaavan/rust-lightning/commits/pr3246.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.09..pr3246.10)):
Addressed @jkczyz comments

Changes:
1. Update `BlindedPathType` documentation.
2. Update code to amend all returned paths by `MessageRouter` to offers, and refund
3. DRY `create_offer_builder`, and `create_refund_builder`

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9124,22 +9124,19 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 			OffersContext::InvoiceRequest { nonce }
 		);
+
+		let offer_builder = OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
```
Just call this `builder`. Any methods on it consume it so it's ok to reuse the name later. Same for `refund_builder`.

shaavan replied:
Done! [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11)
Thanks!

shaavan commented:
Updated from [pr3246.10](https://github.com/shaavan/rust-lightning/commits/pr3246.10) to [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.10..pr3246.11)):
Addressed @jkczyz comments

Changes:
1. Reverted the "Unified Flow" commits, hence re-introducing the two function flows for blinded path creation.
2. Squashed the rest of the commits.
3. Moved BlindedPathType to channelmanager, as it is only used there.

shaavan commented:
Updated from [pr3246.10](https://github.com/shaavan/rust-lightning/commits/pr3246.10) to [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.10..pr3246.11)):
Addressed @jkczyz comments

Changes:
1. Reverted the "Unified Flow" commits, hence re-introducing the two function flows for blinded path creation.
2. Squashed the rest of the commits.
3. Moved BlindedPathType to channelmanager, as it is only used there.
4. Update `DefaultMessageRouter::create_compact_blinded_paths` to only return a single path.

shaavan commented:
Updated from [pr3246.11](https://github.com/shaavan/rust-lightning/commits/pr3246.11) to [pr3246.12](https://github.com/shaavan/rust-lightning/commits/pr3246.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.11..pr3246.12)):

Changes:
1. Rebase on main to resolve conflicts

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -222,6 +222,34 @@ pub enum PendingHTLCRouting {
 	},
 }
 
+/// Represents the types of [`BlindedMessagePath`] that can be created.
```
Could you clarify what sort of interface you want for `create_offer_builder` and `create_refund_builder`? It's not clear to me what you're looking for that will satisfy the three use cases that I outlined in https://github.com/lightningdevkit/rust-lightning/pull/3246#discussion_r1805066862. Are you just asking for a rename or something else?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9123,17 +9152,21 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 
-		let builder = match absolute_expiry {
-			None => builder,
-			Some(absolute_expiry) => builder.absolute_expiry(absolute_expiry),
```
Could you be explicit about what you want the `MessageRouter` interface to look like? Feels like we are going back and forth without much progress...

shaavan replied:
I think this approach makes sense!

Assuming the `scid` lookup is almost free, if `MessageRouter` has access to `MessageForwardNode` and `WhyWeWantABlindedPath`, it can make the right choice between creating a full-length or compact blinded path.

Additionally, if it’s the main decider on path compactness, it could manage the following setting:

```rust
for path in &mut paths {
	path.use_compact_introduction_node(&network_graph);
}
```

This way, it can help avoid creating a compact last hop for full-length paths.

To support the three cases Jeff mentioned in his [comment](https://github.com/lightningdevkit/rust-lightning/pull/3246#discussion_r1805066862), we could consider updating the `create_offer_builder` input fields like so:

```rust
fn create_offer_builder(..., why: Option<WhyWeWantABlindedPath>) -> Offer {
	...
}
```

1. Passing `None` supports creating an offer without a blinded path.
2. Passing `WhyWeWantABlindedPath::ForOffer { offer_expiry_time, is_size_constraint: true, ... }` covers the offer for QR codes.
3. Passing `WhyWeWantABlindedPath::ForOffer { offer_expiry_time, is_size_constraint: false, ... }` covers the unconstrained setting (e.g., an offer on wire).

@TheBlueMatt

I do have a couple of things I’m curious about with this approach:

→ Since `MessageRouter` is a `pub` trait and could be implemented outside of LDK, what if someone uses it in a setting where `scid` lookup for creating `MessageForwardNode` is resource-intensive?

→ Would it be useful for `MessageRouter` to send a note upstream on which type of Blinded Path it ultimately created?

jkczyz replied:
Seems like there would be some duplication with `MessageContext`. And if a new variant is added to `MessageContext`, then one would need to be added to `WhyWeWantABlindedPath`, too. How about:

```rust
pub enum BlindedPathUse {
    Persistent,
    Ephemeral { expiry: Duration },
    ReplyPath,
}
```

That way were aren't tying it strictly to existing BOLT12 use cases.

jkczyz replied:
Oh, but that won't be `absolute_expiry` won't be sufficient for saying no blinded paths. 😕

jkczyz replied:
> Hmm, maybe? I was thinking about the context because someone might want to have a different policy for refunds vs offers, no matter the specific expiry, but maybe we don't need to support that?

It's still possible to differentiate using the passed `MessageContext`.

> Indeed, supporting non-BOLT12 is a bit difficult, though, so maybe the more generic terminology is right.

Yeah, I think I'd prefer that, but we still need to figure out the `ChannelManager` API. We'll need to continue to pass the absolute expiry, but now we need another parameter to indicate whether to include any blinded paths. Using `Option<BlindedPathUse>` won't work since `BlindedPathUse::ReplyPath` isn't applicable.

We could drop that variant in favor of using `Persistent` for both that and long-lived offers/refund given we don't need to use the compact representation for reply paths, even though their lifetime is ephemeral.

jkczyz replied:
> I'm not sure I'm convinced it needs to change. The interface currently allows someone to decide to/not to include a blinded path based on the type of thing we're building, so the question is whether someone is going to want to generate both blinded and non-blinded versions of the same object(s) at the same time in the same node. I can see someone wanting an option to switch it, but that option can be in their route-wrapper, the question is if they will hit races in doing so.

Regarding races, seems we could make methods like `create_offer_builder_using_router` in order to pass a custom router. `DefaultMessageRouter` is cheap to construct, so I can imagine we could also provide methods on it to customize it for one-time use (i.e., by wrapping it). For example:

```rust
let router = DefaultMessageRouter::new(network_graph, entropy_source).without_blinded_paths();
let offer = channel_manager.create_offer_builder_with_router(router).build().unwrap();
```

Or provide a separate `MessageRouter` implementation that returns no paths. Either approach would have equivalent behavior.

> If a user does want this, however, they might also want some different path selection that isn't just "blinded path or not", they might want more/fewer/longer/shorter/compact/non-compact paths based on the specific call. The only way to allow for that, I think, is to pass some kind of opaque "ID" through the `ChannelManager` to the router which the user can map on.

Passing a custom router would be cleaner, IMO, given it can be essentially stateless. `ChannelManager` only needs to hold one for replies. Though arguably a future `OffersMessageFlow` would own that instead. In that world, we could have the `OffersMessageFlow`'s user-provided trait parameterization specify how to create reply routes. It would make it such that `OffersMessageFlow` wouldn't need a direct `MessageRouter` parameterization, though that may not be an entirely necessary thing to do.

jkczyz replied:
> Hmm, not quite visualizing this one but ISTM that we'd want `OffersMessageFlow` to do most of the things `ChannelManager` does today, mostly just moving the code into the new flow manager.

Suppose that `OffersMessageFlow` was parameterized by some `ResponseStrategy` trait (for lack of a better name), which the user must either implement on their own or use a `DefaultResponseStrategy` provided by LDK. The trait could be responsible not only for whether a response should be sent (e.g., if the `InvoiceRequest`'s msat-denominated amount is sufficient for a fiat-denominated offer) but also for how to construct the reply path (i.e., by returning a specific `MessageRouter` as defined by an associated type on the trait).

jkczyz replied:
FWIW, I'm not convinced we need to do what I suggested. Reply path construction doesn't really need much information outside what we'd already provide the `MessageRouter` with the `MessageContext`. I was mostly thinking out loud about how we might remove the `MessageRouter` trait entirely from `OffersMessageFlow`. For creation / initiation uses, the user would provide it directly to whichever method. For the handler, it would need to come from the parameterized `ResponseStrategy`. But that simply moves the parameterization from `OffersMessageFlow` to an associated type on `ResponseStrategy`. So probably doesn't really gain us much.

jkczyz replied:
> Can we just straight up give the `MessageRouter` the `Offer` we're replying to?

`MessageContext` is probably better as it would include any new uses.

shaavan commented:
Updated from [pr3246.12](https://github.com/shaavan/rust-lightning/commits/pr3246.12) to [pr3246.13](https://github.com/shaavan/rust-lightning/commits/pr3246.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.12..pr3246.13)):

Changes:
1. Rebase on main.

shaavan commented:
Updated from [pr3246.13](https://github.com/shaavan/rust-lightning/commits/pr3246.13) to [pr3246.14](https://github.com/shaavan/rust-lightning/commits/pr3246.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.13..pr3246.14)):

Changes:
1. Rebase on #3412 
2. Introduce the "Router-based" approach.
3. With this a user can either pass the `DefaultMessageRouter`, configured to return a particular type of blinded path (or no blinded paths), or can pass a custom message router to the create_{offer/refund}_builder, to return customised blinded paths.

shaavan commented:
@TheBlueMatt, @jkczyz 

A gentle ping! Would love to get your thoughts on this new approach, when you get a moment!
Thanks a lot! :)

shaavan commented:
Updated from [pr3246.14](https://github.com/shaavan/rust-lightning/commits/pr3246.14) to [pr3246.15](https://github.com/shaavan/rust-lightning/commits/pr3246.15) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.14..pr3246.15)):

Changes:
1. Rebase on main, to resolve merge conflicts

shaavan commented:
Updated from [pr3246.15](https://github.com/shaavan/rust-lightning/commits/pr3246.15) to [pr3246.16](https://github.com/shaavan/rust-lightning/commits/pr3246.16) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.15..pr3246.16)):
Addressed @jkczyz comments:

Changes:
1. Introduce `create_{offer, refund}_builder_using_router`, to create blinded paths using custom router.

shaavan commented:
Updated from [pr3246.16](https://github.com/shaavan/rust-lightning/commits/pr3246.16) to [pr3246.17](https://github.com/shaavan/rust-lightning/commits/pr3246.17) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.16..pr3246.17)):

### Changes
- Reverted the rebase from #3412.

### Reasoning
Since the approach in #3412 is still under discussion, I felt it might be better to un-rebase this PR for now. This keeps the thread active and allows us to continue the conversation here in parallel with the ongoing `OffersMessageFlow` work.

Open to thoughts or suggestions, of course!

shaavan commented:
Updated from [pr3246.16](https://github.com/shaavan/rust-lightning/commits/pr3246.16) to [pr3246.17](https://github.com/shaavan/rust-lightning/commits/pr3246.17) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.16..pr3246.17)):

### Changes
- Reverted the rebase from #3412.

**Reasoning**
Since the approach in #3412 is still under discussion, I felt it might be better to un-rebase this PR for now. This keeps the thread active and allows us to continue the conversation here in parallel with the ongoing `OffersMessageFlow` work.

Open to thoughts or suggestions, of course!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2873,9 +2872,11 @@ const MAX_NO_CHANNEL_PEERS: usize = 250;
 /// short-lived, while anything with a greater expiration is considered long-lived.
```
Doesn't look like `MAX_SHORT_LIVED_RELATIVE_EXPIRY` is used anywhere outside of tests now?

shaavan replied:
Thanks for catching it!
Updated it to be test only `const` in [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18)

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2873,9 +2872,11 @@ const MAX_NO_CHANNEL_PEERS: usize = 250;
-/// - [`MessageRouter::create_blinded_paths`] when long-lived.
+/// will included a [`BlindedMessagePath`] created using [`MessageRouter::create_blinded_paths`].
+///
+/// To use compact [`BlindedMessagePath`] one can use [`ChannelManager::create_offer_builder_using_router`]
+/// and [`ChannelManager::create_refund_builder_using_router`] to pass a custom router for compact
+/// blinded path creation.
```
Technically, a user could parameterize `ChannelManager` with `CompactMessageRouter` to achieve the same thing. Though that would mean reply paths (when created from `ChannelManager`, not `OnionMessenger`) would also use the compact format, which while is not wrong is probably not necessary.

shaavan replied:
Makes sense!
Updated the documentation to reflect both the methods in [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18)
Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10197,27 +10198,71 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
+	///
+	/// # Privacy
+	///
+	/// Uses [`MessageRouter`] to construct a [`BlindedMessagePath`] for the offer based on the given
+	/// `absolute_expiry` according to [`MAX_SHORT_LIVED_RELATIVE_EXPIRY`]. See those docs for
```
This is no longer the case.

shaavan replied:
Yep! I think I mixed the documentation while updating!
Thanks for catching it! Fixed in [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18)

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10197,27 +10198,71 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	pub fn create_offer_builder_using_router<ME: MessageRouter>(
+		&$self,
+		router: ME,
+	) -> Result<$builder, Bolt12SemanticError> {
+		let node_id = $self.get_our_node_id();
+		let expanded_key = &$self.inbound_payment_key;
+		let entropy = &*$self.entropy_source;
+		let secp_ctx = &$self.secp_ctx;
+
+		let nonce = Nonce::from_entropy_source(entropy);
+		let context = MessageContext::Offers(OffersContext::InvoiceRequest { nonce });
+
+		let peers = $self.get_peers_for_blinded_path();
+
+		let path = router.create_blinded_paths(node_id, context, peers, secp_ctx)
+			.map_err(|_| Bolt12SemanticError::MissingPaths)?
+			.into_iter().next();
+
+		let mut builder = OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
+			.chain_hash($self.chain_hash);
+
+		if let Some(path) = path {
+			builder = builder.path(path)
+		}
 
 		Ok(builder.into())
 	}
```
Would be nice if we could DRY up the code.

shaavan replied:
Done! Used a closure based approach to DRY up both create offer, and refund builders in [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18)
Thanks for the pointer!

shaavan commented:
Updated from [pr3246.17](https://github.com/shaavan/rust-lightning/commits/pr3246.17) to [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.17..pr3246.18)):
Addressed @jkczyz comments

Changes:
1. Update create_blinded_paths_from_iter() to ignore scid when user wants to create full-length blinded paths.
2. Make `MAX_SHORT_LIVED_RELATIVE_EXPIRY` test only const.
3. Update documentation on ways user can create compact blinded paths.
4. Correct create_offer_builder_using_router documentation.
5. DRY up create_{offer_,refund}_builder.

shaavan commented:
Updated from [pr3246.17](https://github.com/shaavan/rust-lightning/commits/pr3246.17) to [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.17..pr3246.18)):
Addressed @jkczyz comments

Changes:
1. Update create_blinded_paths_from_iter() to ignore scid when user wants to create full-length blinded paths.
2. Make `MAX_SHORT_LIVED_RELATIVE_EXPIRY` test only const.
3. Update documentation on ways user can create compact blinded paths.
4. Correct create_offer_builder_using_router documentation.
5. DRY up create_{offer_,refund}_builder.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -659,78 +639,108 @@ where
-	fn create_compact_blinded_paths<T: secp256k1::Signing + secp256k1::Verification>(
-		&self, recipient: PublicKey, context: MessageContext, peers: Vec<MessageForwardNode>,
-		secp_ctx: &Secp256k1<T>,
+/// A special [`MessageRouter`] that can doesn't route.
```
We should note that this cannot be used to parameterize `ChannelManager` as it wouldn't be able to create reply paths.

shaavan replied:
Done! [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -659,78 +639,108 @@ where
-			secp_ctx,
-			true,
-		)
+/// A [`MessageRouter`] that can only route to a directly connected [`Destination`],
+/// and tries to create compact blinded path for that purpose.
```
We should note that if the caller passes `create_blinded_paths` with `MessageForwardNode`s that do not contain an scid, then the paths will not be compact.

shaavan replied:
Makes sense!
Updated the docs [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19) to properly reflect the best-effort basis nature of `CompactMessageRouter` in [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19)
Thanks a lot for the pointer!

jkczyz commented:
> I'm definitely okay with having methods to override the router used, and like the changes to not decide in `ChannelManager` if a path should be compact or not, but I'm a bit confused because now it seems like we default to doing the wrong thing - if an offer is short-lived we _never_ build a compact blinded path unless the user manually asks for it by passing a `CompactMessageRouter`? That seems like an important regression (unless we think actually its not worth it to have compact BPs in short-lived offers cause most offers are long-lived?).

It could create compact paths by default if `ChannelManager` were parameterized by `CompactMessageRouter`. And besides from having the user explicitly pass `CompactMessageRouter` to a one of the methods, how could it be done without having `ChannelManager` decide?

I guess by passing the `Option<Duration>` through from `ChannelManager` to `MessageRouter`? Not sure it if is even worth having `CompactMessageRouter` then. `NullMessageRouter` may be sufficient to pass as an override. Don't have strong opinion, though it would be nice if `create_offer_builder` didn't need an optional parameter.

shaavan commented:
Updated from [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18) to [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.18..pr3246.19)):
Addressed @jkczyz comments

Changes:
1. Expanded `NullMessageRouter`, and `CompactMessageRouter` document.

shaavan commented:
Updated from [pr3246.18](https://github.com/shaavan/rust-lightning/commits/pr3246.18) to [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.18..pr3246.19)):
Addressed @jkczyz comments

Changes:
- Expanded `NullMessageRouter`, and `CompactMessageRouter` document.

shaavan commented:
I see the point you're suggesting, Matt, and I agree there’s some convenience in having the router automatically decide the blinded path type based on `Option<Duration>`. But I think having the user explicitly call the offer builder with a `CompactMessageRouter` is just as convenient—and arguably cleaner.

The current approach has the benefit of being **explicit and predictable**. By default, we *always* generate a full-length path, providing a consistent baseline across all offers. If a user wants a compact path—for example, for short-lived offers—they can opt into it clearly via the router they pass in. Conversely, if the user parameterizes `ChannelManager` with a `CompactMessageRouter`, they’ll consistently get compact paths—unless they manually override that choice.

This trades away automatic decision-making by the router in favor of something I believe is more valuable in the long run: **user awareness and control**. It also avoids introducing heuristics that could surprise users unless they’re quite familiar with the internal logic.

In that sense, prioritizing clear defaults and explicit opt-in behavior, I believe, makes the design more intuitive and easier to reason about over time.

Curious to hear your thoughts—thanks!

jkczyz commented:
> I see the point you're suggesting, Matt, and I agree there’s some convenience in having the router automatically decide the blinded path type based on `Option<Duration>`. But I think having the user explicitly call the offer builder with a `CompactMessageRouter` is just as convenient—and arguably cleaner.

Regarding this point, I tend to agree that removing the `Option<Duration>` from methods like `create_offer_builder` is more desirable. The existing behavior uses this to set the expiry and provide additional information for creating the path. But upon returning an `OfferBuilder`, the user can then override the expiry with a longer one and be left with a long-lived offer using compact blinded paths.

shaavan commented:
Updated from [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19) to [pr3246.20](https://github.com/shaavan/rust-lightning/commits/pr3246.20) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.19..pr3246.20)):
Addressed @TheBlueMatt comments

Changes:
1. Update `DefaultMessageRouter` to create Compact `BlindedMessagePath` by default.
2. When a user wants to create full-length blinded path, they can parameterise using the `NodeIdMessageRouter`

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -706,6 +706,114 @@ where
+/// A special [`MessageRouter`] implementation that performs no routing.
+///
+/// # Note
+/// [`NullMessageRouter`] **must not** be used with [`ChannelManager`] as a parameter.
```
Well, it can be used with `ChannelManager` methods, but I guess shouldn't be used as its _type_ parameter.

shaavan replied:
Makes sense! I have update the docs to make it explicit, with example too.
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -706,6 +706,114 @@ where
+	}
+}
+
+/// A special [`MessageRouter`] implementation that performs no routing.
```
The purpose of this isn't for no routing, but rather to create offers/refunds without blinded paths.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -706,6 +706,114 @@ where
+/// # Note
+/// [`NullMessageRouter`] **must not** be used with [`ChannelManager`] as a parameter.
+///
+/// # Reason
+/// [`ChannelManager`] requires a functioning [`MessageRouter`] to create blinded paths,
+/// which are necessary for constructing reply paths in onion message communication.
```
This sentence belongs in the previous section. No need for a separate section.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -3374,6 +3387,33 @@ pub fn create_node_cfgs_with_persisters<'a>(node_count: usize, chanmon_cfgs: &'a
 	nodes
 }
 
+pub fn create_node_cfgs<'a>(node_count: usize, chanmon_cfgs: &'a Vec<TestChanMonCfg>) -> Vec<NodeCfg<'a>> {
+	let persisters = chanmon_cfgs.iter().map(|c| &c.persister).collect();
+	create_node_cfgs_internal(node_count, chanmon_cfgs, persisters, |graph, keys_manager| {
+		test_utils::TestMessageRouter::new_default(graph, keys_manager)
+	})
+}
+
+pub fn create_node_cfgs_with_persisters<'a>(
+	node_count: usize,
+	chanmon_cfgs: &'a Vec<TestChanMonCfg>,
+	persisters: Vec<&'a impl test_utils::SyncPersist>,
+) -> Vec<NodeCfg<'a>> {
+	create_node_cfgs_internal(node_count, chanmon_cfgs, persisters, |graph, keys_manager| {
+		test_utils::TestMessageRouter::new_default(graph, keys_manager)
+	})
+}
+
+pub fn create_node_cfgs_with_node_id_message_router<'a>(
+	node_count: usize,
+	chanmon_cfgs: &'a Vec<TestChanMonCfg>,
+) -> Vec<NodeCfg<'a>> {
+	let persisters = chanmon_cfgs.iter().map(|c| &c.persister).collect();
+	create_node_cfgs_internal(node_count, chanmon_cfgs, persisters, |graph, keys_manager| {
+		test_utils::TestMessageRouter::new_node_id_router(graph, keys_manager)
+	})
+}
```
nit: For these, you don't need to pass a closure because the constructors take the same arguments as the expected closure type. So just pass `test_utils::TestMessageRouter::new_default`, for example.

shaavan replied:
Ah! Thanks for the pointer!
Updated in [pr3246.23](https://github.com/shaavan/rust-lightning/commits/pr3246.23)

jkczyz commented on lightning/src/util/test_utils.rs:
```diff
@@ -311,41 +311,74 @@ impl<'a> Drop for TestRouter<'a> {
-		&'a TestLogger,
-		&'a TestKeysInterface,
-	>,
+pub enum TestMessageRouter<'a> {
+	Default {
+		inner: DefaultMessageRouter<
+			Arc<NetworkGraph<&'a TestLogger>>,
+			&'a TestLogger,
+			&'a TestKeysInterface,
+		>,
+	},
+
+	NodeId {
+		inner: NodeIdMessageRouter<
+			Arc<NetworkGraph<&'a TestLogger>>,
+			&'a TestLogger,
+			&'a TestKeysInterface,
+		>,
+	},
 }
```
No need for explicit `inner`. Just use tuple struct syntax for each variant, e.g., `Default(DefaultMessageRouter<...>)`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10852,10 +10851,9 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
-	/// [`MessageRouter`].
+	/// Uses the [`ChannelManager`]'s [`MessageRouter`] to construct a [`BlindedMessagePath`] for
+	/// the offer. See those docs for privacy implications as well as those of the parameterized
+	/// [`Router`], which implements [`MessageRouter`].
```
`Router` no longer implements `MessageRouter`. This may have been missed when we decoupled them. Could you add a commit dropping that part?

shaavan replied:
Added a separate fixup commit updating the docs in [pr3246.23](https://github.com/shaavan/rust-lightning/commits/pr3246.23)

jkczyz commented on lightning/src/offers/flow.rs:
```diff
@@ -449,61 +449,107 @@ where
+			PublicKey,
+			MessageContext,
+			&secp256k1::Secp256k1<secp256k1::All>,
+		) -> Result<Vec<BlindedMessagePath>, Bolt12SemanticError>,
```
Can this return an `Iterator` so we don't need to `collect` in two instances below?

shaavan replied:
Makes sense! Updated in [pr3246.23](https://github.com/shaavan/rust-lightning/commits/pr3246.23) to return iterator, instead of `vec`. Thanks!

jkczyz commented on lightning/src/util/test_utils.rs:
```diff
@@ -311,41 +311,69 @@ impl<'a> Drop for TestRouter<'a> {
-		&'a TestLogger,
-		&'a TestKeysInterface,
-	>,
+pub enum TestMessageRouter<'a> {
```
Instead of introducing this enum, can each test that needs to a different `MessageRouter` simply override `NodeCfg::message_router` before calling `create_node_chanmgrs`?

shaavan replied:
Thanks a lot for the suggestion!

I actually experimented with overriding NodeCfg::message_router before going with the enum approach. The main challenge I ran into was that NodeCfg is currently configured with a concrete TestMessageRouter type. Switching it to be generic over the MessageRouter trait would require parameterizing NodeCfg itself, which would ripple into a lot of unrelated parts of the test code.

To avoid that complexity, I chose to modify TestMessageRouter into an enum instead. It felt like a cleaner tradeoff that kept the impact localized.

Let me know if this reasoning makes sense to you. Thanks so much for the pointer!

shaavan replied:
Thanks a lot for the suggestion!

I actually experimented with overriding `NodeCfg::message_router` before going with the enum approach. The main challenge I ran into was that `NodeCfg` is currently configured with a concrete `TestMessageRouter` type. Switching it to be generic over the `MessageRouter` trait would require parameterizing `NodeCfg` itself, which would ripple into a lot of unrelated parts of the test code.

To avoid that complexity, I chose to modify `TestMessageRouter` into an enum instead. It felt like a cleaner tradeoff that kept the impact localized.

Let me know if this reasoning makes sense to you. Thanks so much for the pointer!

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -406,7 +416,7 @@ fn creates_short_lived_offer() {
 fn creates_long_lived_offer() {
 	let chanmon_cfgs = create_chanmon_cfgs(2);
-	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_cfgs = create_node_cfgs_with_node_id_message_router(2, &chanmon_cfgs);
```
We can use `create_offer_builder_using_router` instead.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -470,7 +480,7 @@ fn creates_short_lived_refund() {
 fn creates_long_lived_refund() {
 	let chanmon_cfgs = create_chanmon_cfgs(2);
-	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_cfgs = create_node_cfgs_with_node_id_message_router(2, &chanmon_cfgs);
```
Similar.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10918,12 +10916,46 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
+	/// Creates an [`OfferBuilder`] such that the [`Offer`] it builds is recognized by the
+	/// [`ChannelManager`] when handling [`InvoiceRequest`] messages for the offer.
+	///
+	/// # Privacy
+	///
+	/// Constructs a [`BlindedMessagePath`] for the offer using a custom [`MessageRouter`].
+	/// Users can implement a custom [`MessageRouter`] to define properties of the
+	/// [`BlindedMessagePath`] as required or opt not to create any `BlindedMessagePath`.
+	///
+	/// Also, uses a derived signing pubkey in the offer for recipient privacy.
+	///
+	/// # Limitations
+	///
+	/// See [`OffersMessageFlow::create_offer_builder`] for limitations on the offer builder.
+	///
+	/// # Errors
+	///
+	/// Errors if the provided [`MessageRouter`] is unable to create a blinded path for the offer.
+	///
+	/// [`BlindedMessagePath`]: crate::blinded_path::message::BlindedMessagePath
+	/// [`Offer`]: crate::offers::offer::Offer
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
```
Instead of duplicating the docs, could you instead reference `ChannelManager::create_offer_builder` docs, noting the differences. Similarly in `OffersMessageFlow`.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10901,9 +10900,8 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
-	/// Uses [`MessageRouter`] to construct a [`BlindedMessagePath`] for the offer based on the given
-	/// `absolute_expiry` according to [`MAX_SHORT_LIVED_RELATIVE_EXPIRY`]. See those docs for
-	/// privacy implications.
+	/// Uses the [`ChannelManager`]'s [`MessageRouter`] to construct a [`BlindedMessagePath`] for
```
Won't this use `OffersMessageFlow`'s message router? Seems like `ChannelManager` no longer has one as it is being passed through to `OffersMessageFlow`. Since we are only creating the flow internally now, perhaps we should word this as:

```rust
/// Uses the [`MessageRouter`] passed to [`ChannelManager`]'s constructor to ...
```

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11024,6 +11024,72 @@ macro_rules! create_refund_builder { ($self: ident, $builder: ty) => {
 		Ok(builder.into())
 	}
+
+	/// Creates a [`RefundBuilder`] such that the [`Refund`] it builds is recognized by the
```
Similarly regarding repeating the docs.

shaavan commented:
Updated from [pr3246.24](https://github.com/shaavan/rust-lightning/commits/pr3246.24) to [pr3246.25](https://github.com/shaavan/rust-lightning/commits/pr3246.25) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.24..pr3246.25)):
Addressed @jkczyz comments

Changes:
1. Removed redundant code snippet from test.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1426,7 +1464,10 @@ where
 			message_recipients
 				.iter()
 				.filter(|(_, peer)| matches!(peer, OnionMessageRecipient::ConnectedPeer(_)))
-				.map(|(node_id, _)| *node_id)
+				.map(|(node_id, _)| MessageForwardNode {
+					node_id: *node_id,
+					short_channel_id: None,
+				})
```
@TheBlueMatt This means even if we are configured with a `DefaultMessageRouter`, one of the hops in the created reply path won't be compact. Not sure if there is much we could do about this, but I just want to call it out. This affects three uses:
- `OnionMessenger::send_onion_message` for stand-alone onion message sending
- `OnionMessenger::handle_onion_message_response` or async response handling
- `OnionMessenger::next_onion_message_for_peer` for sync response handling

Prior to this change we'd always use non-compact paths when creating reply paths. But now it is based on the configured `MessageRouter` except here where we don't have an SCID for the peer.

jkczyz replied:
I suppose we could parameterize `OnionMessenger` with a `NetworkGraph` instead of a `MessageRouter` such that we can create a `NodeIdMessageRouter` internally? This would at least make those reply path hops consistent. However, when using `ChannelMangaer::pay_for_offer` we'll still make reply paths using (exclusively) compact hops.

jkczyz reviewed (commented):
Code LGTM. Just one comment directed at @TheBlueMatt.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -501,41 +501,28 @@ pub trait MessageRouter {
+///
+/// Compact blinded paths use short channel IDs (SCIDs) instead of pubkeys, resulting in smaller
+/// serialization. This is particularly useful when encoding data into space-constrained formats
+/// such as QR codes. The SCID is communicated via a [`MessageForwardNode`], but may be `None`
```
Hmmm... *with* splicing, the SCID still changes -- it's the `ChannelId` that remains the same. I suppose if we us an SCID alias, then it wouldn't be a problem. But wouldn't that only be relevant for only the last hop in the path? We'd use real SCIDs from the `NetworkGraph` for the other hops.

shaavan commented:
Updated from [pr3246.25](https://github.com/shaavan/rust-lightning/commits/pr3246.25) to [pr3246.26](https://github.com/shaavan/rust-lightning/commits/pr3246.26) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.25..pr3246.26)):
Addressed @jkczyz, @TheBlueMatt comments

Changes:
1. Expand the comment documenting the limitation of Compact blinded paths.
2. Fix indentation.

shaavan commented:
Updated from [pr3246.26](https://github.com/shaavan/rust-lightning/commits/pr3246.26) to [pr3246.27](https://github.com/shaavan/rust-lightning/commits/pr3246.27) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3246.26..pr3246.27)):

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
> Went ahead and did the followups since I was here anyway https://github.com/lightningdevkit/rust-lightning/pull/3959

Matt on fire! 🔥
Thanks a lot for all the reviews, and catching the cleanups! ✨
